README.md
# Component Help Installer

A Delphi Pascal Windows Program to help install OpenHelp compatible WinHelp component help files into the IDE for Delphi versions 3 to 7.

> ***This project is archived and is no longer being developed. No support is available. Bug reports, feature requests and pull requests are not being accepted.***

For more information about the program please see the [Component Help Installer](https://delphidabbler.com/software/chi) page on [delphidabbler.com](https://delphidabbler.com).

For information about using the program see the WinHelp help file or `Src/CHI/Help/CHI.rtf`.

## Final Release

The final release of the program is [v3.0.0](https://github.com/ddab-archive/chi/releases/tag/v3.0.0-FINAL).

## Source Code

The source code was not originally developed under version control. Eventually the source code was moved to a Subversion repository. A summary of changes made before moving to version control can be found in [`Docs/PreSVNHistory.txt`](https://github.com/ddab-archive/chi/blob/main/Docs/PreSVNHistory.txt). The remainder of the history was recorded in the Subversion repo, a dump of which has been uploaded to `SVN/chi-svndump.zip` [[download](https://github.com/ddab-archive/chi/raw/main/SVN/chi-svndump.zip)].

### Commits since the last release

The `main` branch of this project differs from the final release in two ways:

1. This `README.md` file has been added for display by GitHub.
2. `8queens-svndump.zip` was uploaded to the `SVN` directory, as noted above.

The source code as it was at the final release can be viewed on GitHub by switching to the [`v3.0.0-FINAL`](https://github.com/ddab-archive/chi/releases/tag/v3.0.0-FINAL) tag.

## License

The executable code is covered by the EULA described in `Docs/License.rtf`.

Much of the source code released under the [Mozilla Public License v1.1](https://www.mozilla.org/MPL/1.1/).


Docs/ChangeLog.txt
; ------------------------------------------------------------------------------
; Change Log for Component Help Installer
; $Rev: 73 $
; $Date: 2010-05-10 17:01:47 +0100 (Mon, 10 May 2010) $
; ------------------------------------------------------------------------------

Release 3.0 of 10 May 2010
CHI.exe v3.0 (build 10) of 10 May 2010
CHInst.exe v1.1 (build 9) of 10 May 2010
CHInstLib.dll v1.2 (build 9) of 10 May 2010
+ Removed Delphi IDE add-in expert from product: it now contains stand alone GUI and command line apps only.
+ Subsumed former PJCHI.dll functionality into CHI.exe.
+ Removed GUI program's options dialog. It was used to configure IDE add-in.
+ Completely revised GUI program's about box.
+ Added manifest resources to set required admin execution level on Vista and later.
+ Removed online registration and email author options from GUI program.
+ Removed splash screen from GUI program and associated nags.
+ Updated error messages when program runs without correct privileges.
+ Programs no longer access or update the registry.
+ New GUI installer replaces old command line installer. Support for Vista and later added and support for Win 9x removed.
+ All .exe and .dll files must now be in same directory.
+ Changed to use updated DelphiDabbler System Info library code.
+ Removed tool used to convert CHI v1 files to v2 format.
+ Restructured source code and removed unused files.
+ Updated help file with changes.
+ New EULA.
+ Compiled with Delphi 7 instead of Delphi 4.

Release 2.2.1 of 12 December 2005
CHI.exe v2.0 (build 9) of 12 Dec 2005
CHIExp.dll v2.0 (build 10) of 12 Dec 2005
CHInst.exe v1.0.1 (build 8) of 12 Dec 2005
CHInstLib.dll v1.1.1 (build 8) of 12 Dec 2005
PJCHI.dll v2.1 (build 11) of 12 Dec 2005
+ Fixed bug where program crashes under Delphi 6 and 7 when no .als file is present.

Release 2.2 of 13 November 2005
CHI.exe v2.0 (build 8) of 13 Nov 2005
CHIExp.dll v2.0 (build 9) of 13 Nov 2005
CHInst.exe v1.0.1 (build 7) of 13 Nov 2005
CHInstLib.dll v1.1 (build 7) of 13 Nov 2005
PJCHI.dll v2.1 (build 10) of 13 Nov 2005
+ Added support for installing .als files in Delphi 6 and 7.
+ Updated help file accordingly.
+ Changed to Mozilla Public License.

Release 2.1 of 17 April 2004
CHI.exe v2.0 (build 7) of 17 Apr 2004
CHIExp.dll v2.0 (build 8) of 17 Apr 2004
CHInst.exe v1.0.1 (build 6) of 17 Apr 2004
CHInstLib.dll v1.0.1 (build 6) of 17 Apr 2004
PJCHI.dll v2.1 (build 9) of 17 Apr 2004
+ The installed path of the selected help file is now displayed on the "Remove" page of the GUI program and expert.
+ Fixed bug that prevented help file from being unregistered with Delphi if it is not registered with Windows.

Release 2.0.1 of 22 August 2003
CHI.exe v2.0 (build 6) of 22 Aug 2003
CHIExp.dll v2.0 (build 7) of 22 Aug 2003
CHInst.exe v1.0.1 (build 5) of 22 Aug 2003
CHInstLib.dll v1.0.1 (build 5) of 22 Aug 2003
PJCHI.dll v2.0.1 (build 8) of 22 Aug 2003
+ Fixed bug in GUI application that was preventing program from refusing to run when Delphi is not installed on the computer.
+ Also fixed installer bug when run on systems with no Delphi installation.

Release 2.0 of 20 July 2003
CHI.exe v2.0 of 20 Jul 2003
CHIExp.dll v2.0 of 19 Jul 2003
CHInst.exe v1.0 of 20 Jul 2003
CHInstLib.dll v1.0 of 19 Jul 2003
PJCHI.dll v2.0 of 20 Jul 2003
+ Delphi expert is now contained in a DLL which can be installed into Delphi IDE automatically. This replaces the source code unit (PJCHIExpert.pas) that previously needed to be compiled into a design time package within Delphi.
+ Added dialog box to configure which versions of Delphi the expert is installed into. This dialog can be accessed from the stand-alone application GUI. It can also be displayed from the command line and is used by the install program.
+ Removed link to email comments from hyperlink on main GUI application and add-in expert main window - this is now a menu item accessed from the Help button.
+ Modified command line application so that WinHelp topics can be accessed from it.
+ Some minor bug fixes.
+ Expanded help file re new features.

Release 2.0 (Beta 2) of 24 May 2003
CHI.exe v1.1 (Beta 2) of 24 May 2003
CHInst.exe v0.9 (Beta 2) of 24 May 2003
CHInstLib.dll v0.9 (Beta 2) of 24 May 2003
PJCHI.dll v1.99 (Beta 2) of 24 May 2003
PJCHIExpert.pas v1.4 of 24 May 2003
+ Revised method used to record help file information in Delphi OpenHelp files. The DLL now updates Delphi's index and link files rather than using its own config file. This new method works with later versions of Delphi. A custom config file is now used only to maintain a list of files installed by CHI.
+ Various WinHelp generated housekeeping files are now deleted whenever OpenHelp is updated. This ensures changes take effect immediately.
+ Changed PJSoft branding to DelphiDabbler.
+ Improved program registration function.

Release 2.0 (Beta 1 Patch 1) of 13 January 2002
CHI.exe v1.1 (Beta 1 Patch 1) of 13 Jan 2002
CHInst.exe v0.9 (Beta 1 Patch 1) of 13 Jan 2002
CHInstLib.dll v0.9 (Beta 1 Patch 1) of 13 Jan 2002
PJCHI.dll v1.99 Beta 1 of Patch 1 of 13 Jan 2002
PJCHIExpert.pas v1.3 of 13 Jan 2002
+ Fixes "DLL not found" bug in Delphi expert.

Release 2.0 (Beta 1) of 06 January 2002
CHI.exe v1.1 (Beta 1) of 06 Jan 2002
CHInst.exe v0.9 (Beta 1) of 06 Jan 2002
CHInstLib.dll v0.9 (Beta 1) of 06 Jan 2002
PJCHI.dll v1.99 (Beta 1) of 06 Jan 2002
PJCHIExpert.pas v1.2 of 06 Jan 2002
+ First Beta of new release that aims to be compatible with Delphi 3 to 6.
+ Introduces command line version of the application.
+ New installation engine using different technique to release 1.

Release 1.1.1 of 23 June 2001
CHI.exe v1.0 of 06 Apr 2001
PJCHI.dll v1.1.1 of 23 Jun 2001
PJCHIExpert.pas v1.1 of 23 Jun 2001
+ Fixed bug that was preventing expert and application from detecting Delphi on locales not using '.' as decimal point.

Release 1.1 of 09 June 2001

Src/CHInst/UReporter.pas
{
 * UReporter.pas
 *
 * Defines object used to display information about program on console, and to
 * detect user input.
 *
 * $Rev: 42 $
 * $Date: 2010-04-09 17:25:16 +0100 (Fri, 09 Apr 2010) $
 *
 * ***** BEGIN LICENSE BLOCK *****
 *
 * Version: MPL 1.1
 *
 * The contents of this file are subject to the Mozilla Public License Version
 * 1.1 (the "License"); you may not use this file except in compliance with the
 * License. You may obtain a copy of the License at http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License for
 * the specific language governing rights and limitations under the License.
 *
 * The Original Code is UReporter.pas.
 * 
 * The Initial Developer of the Original Code is Peter Johnson
 * (http://www.delphidabbler.com/).
 * 
 * Portions created by the Initial Developer are Copyright (C) 2001-2010 Peter
 * Johnson. All Rights Reserved.
 *
 * Contributor(s):
 *   NONE
 *
 * ***** END LICENSE BLOCK *****
}


unit UReporter;


interface


type

  {
  TReporter:
    Object used to display information about program on console, and to detect
    user input. Output can be toggled on and off by using Quiet property.
  }
  TReporter = class(TObject)
  private
    fQuiet: Boolean;
      {Flag true if quiet (i.e. no) output}
    fSignedOn: Boolean;
      {Flag true if SignOn method has been called}
    procedure Say(const Line: string);
      {Outputs given line of text on console. No action is taken if Quiet
      property is false}
  public
    procedure SignOn;
      {Outputs sign-on message on console: sign-on message is only displayed
      once regardless of how many times this method is called}
    procedure SignOff;
      {Outputs prompt on console and waits for user to press return. This is not
      done if Quiet property is false}
    procedure DisplayUsage;
      {Display text on console informing user of program's usage, providing
      Quiet property is false}
    procedure DisplayHelp;
      {Displays detailed help on using the program on the console - providing
      Quiet property is false}
    procedure DisplayError(const Msg: string);
      {Displays the given error message on console, ensuring that program has
      signed on first. Also displays program usage. No action taken if Quiet
      property is false}
    procedure DisplayInstallActions(const HelpSourceFile, HelpDestFile,
      Description: string; const DelphiVer: Integer);
      {Displays the actions the program is about to perform on installation:
      info displayed depends on parameters. Nothing is displayed if Quiet
      property is false}
    procedure DisplayUninstallActions(const HelpSourceFile: string;
      const DeleteFile, UnRegister: Boolean; const DelphiVer: Integer);
      {Displays the actions the program is about to perform on uninstallation:
      info displayed depends on parameters. Nothing is displayed if Quiet
      property is false}
    property Quiet: Boolean read fQuiet write fQuiet default False;
      {When this property is true the reporter displays messages on the console
      and reads user input from it. When Quiet is false the reporter ignores
      all display and input requests}
  end;


implementation


{ TReporter }

procedure TReporter.DisplayError(const Msg: string);
  {Displays the given error message on console, ensuring that program has signed
  on first. Also displays program usage. No action taken if Quiet property is
  false}
begin
  // Sign on (if required) and display error message
  SignOn;
  Say('Error: ' + Msg);
  Say('');
  // Display usage and info about getting detailed help
  DisplayUsage;
  Say('Use CHInst -? to display help.');
  Say('');
end;

procedure TReporter.DisplayHelp;
  {Displays detailed help on using the program on the console - providing Quiet
  property is false}
begin
  DisplayUsage;
  Say('-c  copy help file to location specified by next parameter');
  Say('-d  description of help file specified by next parameter');
  Say('-h  help: displays the Windows help file at the Command Line application'
    + ' topic');
  Say('-i  install: expects install file name as next parameter');
  Say('-q  quiet mode: no screen output');
  Say('-r  remove help file from registry when uninstalling');
  Say('-u  uninstall: expects uninstall file name as next parameter');
  Say('-vX Delphi version: X=3..7 (default: -v4)');
  Say('-x  delete help file when uninstalling');
  Say('-?  help: displays this usage information - following parameters are '
    + 'ignored');
  Say('');
  Say('Switches are not case sensitive and can be preceded by '
    + '''/'' or ''-''.');
  Say('Those switches following help file name can occur in any order.');
  Say('$(ROOT) or $(HELP) in file paths are replaced by Delphi''s root and '
    + 'help paths.');
  Say('');
end;

procedure TReporter.DisplayInstallActions(const HelpSourceFile,
  HelpDestFile, Description: string; const DelphiVer: Integer);
  {Displays the actions the program is about to perform on installation: info
  displayed depends on parameters. Nothing is displayed if Quiet property is
  false}
begin
  if HelpDestFile = '' then
    // We have no destination file, so we install source file
    Say('Installing ' + HelpSourceFile)
  else
  begin
    // We have a destination file so we copy source and install destination file
    Say('Copying ' + HelpSourceFile + ' to ' + HelpDestFile);
    Say('Installing ' + HelpDestFile);
  end;
  // We report Delphi version number
  Say('Into Delphi version ' + Chr(DelphiVer + Ord('0')));
  // Report any provided description or say we have no description
  if Description <> '' then
    Say('Help file description is "' + Description + '"')
  else
    Say('Help file has no description');
  Say('');
end;

procedure TReporter.DisplayUninstallActions(const HelpSourceFile: string;
  const DeleteFile, UnRegister: Boolean; const DelphiVer: Integer);
  {Displays the actions the program is about to perform on uninstallation: info
  displayed depends on parameters. Nothing is displayed if Quiet property is
  false}
begin
  // Say which file we're uninstalling, and from which version of Delphi
  Say('Uninstalling ' + HelpSourceFile);
  Say('From Delphi version ' + Chr(DelphiVer + Ord('0')));
  Say('Help file:');
  // Note if we are deleting file
  if DeleteFile then
    Say('  will be deleted')
  else
    Say('  will not be deleted');
  // Note if file is to be un-registered with Windows
  if UnRegister then
    Say('  will be unregistered with Windows')
  else
    Say('  will not be unregistered with Windows');
  Say('');
end;

procedure TReporter.DisplayUsage;
  {Display text on console informing user of program's usage, providing Quiet
  property is false}
begin
  Say('Usage is:');
  Say('  CHInst [-i] help_file [-vX] '
    + '[-c copy_help_file] [-d description] [-q]');
  Say('  CHInst -u help_file [-vX] [-r] [-x] [-q]');
  Say('  CHInst -h | -?');
  Say('');
end;

procedure TReporter.Say(const Line: string);
  {Outputs given line of text on console. No action is taken if Quiet property
  is false}
begin
  if not fQuiet then
    WriteLn(Line);
end;

procedure TReporter.SignOff;
  {Outputs prompt on console and waits for user to press return. This is not
  done if Quiet property is false}
begin
  if not Quiet then
  begin
    Write('Press Enter to end');
    ReadLn;
  end;
end;

procedure TReporter.SignOn;
  {Outputs sign-on message on console: sign-on message is only displayed once
  regardless of how many times this method is called}
begin
  // Check if we have done this before
  if not fSignedOn then
  begin
    Say('DelphiDabbler: Component Help Installer');
    Say('Command Line Version');
    Say('Copyright (c) 2001-2003 P.D.Johnson, Wales, UK');
    Say('----------------------------------------------');
    // Record that we've now output the message
    fSignedOn := True;
  end;
end;

end.

